presto-main/src/main/java/com/facebook/presto/memory/ClusterMemoryManager.java

```diff
@@ -112,10 +134,39 @@ public void process(Iterable<QueryExecution> queries)
             if (bytes > queryMemoryLimit) {
                 DataSize maxMemory = succinctDataSize(queryMemoryLimit, BYTE);
                 query.fail(exceededGlobalLimit(maxMemory));
+                queryKilled = true;
             }
         }
         clusterMemoryUsageBytes.set(totalBytes);
 
+        boolean shouldKillQuery = nanosSince(lastTimeNotOutOfMemory).compareTo(killOnOutOfMemoryDelay) == 1 && isClusterOutOfMemory();
+        boolean lastKilledQueryIsGone = lastKilledQuery == null;
+
+        if (lastKilledQuery != null) {
+            ClusterMemoryPool generalPool = pools.get(GENERAL_POOL);
+            if (generalPool != null) {
+                lastKilledQueryIsGone = generalPool.getQueryMemoryReservations().containsKey(lastKilledQuery);
+            }
+        }
+
+        if (shouldKillQuery && lastKilledQueryIsGone && !queryKilled && killOnOutOfMemory) {
```

Comment:
maybe put `killOnOutOfMemory` first

Actually, why not wrap this whole section in `if (killOnOutOfMemory)`